libnmstate/nm/active_connection.py

```diff
@@ -65,114 +61,83 @@ def deactivate(self):
 
         For software devices, deactivation removes the devices from the kernel.
         """
-        self._mainloop.push_action(self._safe_deactivate_async)
-
-    def _safe_deactivate_async(self):
         act_connection = self._nmdev.get_active_connection()
         if not act_connection or act_connection.props.state in (
             NM.ActiveConnectionState.DEACTIVATING,
             NM.ActiveConnectionState.DEACTIVATED,
         ):
-            # Nothing left to do here, call the next action.
-            self._mainloop.execute_next_action()
             return
 
-        user_data = None
-        act_connection.connect(
-            NM_AC_STATE_CHANGED_SIGNAL, self._wait_state_changed_callback,
+        if self._act_con != act_connection:
+            raise NmstateLibnmError(
+                "When deactivating active connection, the newly get "
+                f"NM.ActiveConnection {act_connection}"
+                f"is different from original request: {self._act_con}"
+            )
+
+        action = f"Deactivate profile: {self.devname}"
```

Comment:
This seems to be acting as an action identifier, which is nice. The risk I see here is that it is not managed, i.e. some other action may use the same naming. You cannot expect someone else that adds an action in the future to go over all existing actions, not it is listen in one place from him to do so.
The chances for this to occur are very rare, but so are races, deadlocks, memory corruptions and other friends.

Leaving it up to you to decide how to proceed.

Reply:
We only have 6 async actions. We can worry about overlaps when we have 10+.
 * Delete device
 * Add profile
 * Modify profile
 * Delete profile
 * Activate profile
 * Deactivate profile